kafka/producer_metric_test.go

```diff
@@ -0,0 +1,149 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kafka
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	"github.com/elastic/apm-data/model"
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/codec/json"
+)
+
+func TestProducerMetrics_deadlineExceeded(t *testing.T) {
+	producer, rdr := setupTestProducer(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 0)
+	defer cancel()
+
+	err := producer.ProcessBatch(ctx, &model.Batch{
+		model.APMEvent{Transaction: &model.Transaction{ID: "1"}},
+		model.APMEvent{Transaction: &model.Transaction{ID: "2"}},
+		model.APMEvent{Span: &model.Span{ID: "3"}},
+	})
+	assert.NoError(t, err)
+
+	var rm metricdata.ResourceMetrics
+	assert.NoError(t, rdr.Collect(context.Background(), &rm))
+
+	metrics := extractMetrics(t, rm.ScopeMetrics)
+	assert.Len(t, metrics, 1)
+	metric := metrics[0]
+
+	want := metricdata.Metrics{
+		Name: "producer.messages.error",
+		Data: metricdata.Sum[int64]{
+			DataPoints: []metricdata.DataPoint[int64]{
+				{Value: 3},
+			},
+		},
+	}
+
+	testMetric(t, want, metric)
+}
+
+func TestProducerMetrics_produced(t *testing.T) {
+	producer, rdr := setupTestProducer(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	err := producer.ProcessBatch(ctx, &model.Batch{
+		model.APMEvent{Transaction: &model.Transaction{ID: "1"}},
+		model.APMEvent{Transaction: &model.Transaction{ID: "2"}},
+		model.APMEvent{Span: &model.Span{ID: "3"}},
+	})
+	assert.NoError(t, err)
+
+	var rm metricdata.ResourceMetrics
+	assert.NoError(t, rdr.Collect(context.Background(), &rm))
+
+	metrics := extractMetrics(t, rm.ScopeMetrics)
+	assert.Len(t, metrics, 1)
+	metric := metrics[0]
+
+	want := metricdata.Metrics{
+		Name: "producer.messages.produced",
+		Data: metricdata.Sum[int64]{
+			DataPoints: []metricdata.DataPoint[int64]{
+				{Value: 3},
+			},
+		},
+	}
+
+	testMetric(t, want, metric)
+}
+
+func extractMetrics(t *testing.T, sm []metricdata.ScopeMetrics) []metricdata.Metrics {
+	t.Helper()
+
+	for _, m := range sm {
+		if m.Scope.Name == instrumentName {
+			return m.Metrics
+		}
+	}
+	t.Fatal("unable to find metrics for", instrumentName)
+	return []metricdata.Metrics{}
+}
+
+func testMetric(t *testing.T, want metricdata.Metrics, actual metricdata.Metrics) {
+	t.Helper()
+
+	assert.Equal(t, want.Name, actual.Name)
+
+	wantValue := want.Data.(metricdata.Sum[int64]).DataPoints[0].Value
+	metricValue := actual.Data.(metricdata.Sum[int64]).DataPoints[0].Value
+	assert.Equal(t, wantValue, metricValue)
+}
```

Comment:
Rather than using manual assertions, did you consider using [`go.opentelemetry.io/otel/sdk/metric/metricdata/metricdatatest`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric/metricdata/metricdatatest)? It seems more complete and it's easier to also assert dimensions. The current assertion completely dismisses dimensions.

Reply:
Given that the test was only targetting name and value using the entire metricdatatest was overkill and producing some issues, as there were fields conflicting then performing the equality check. I worked around that with `go-cmp` but in the end this was simpler and achieving the same goal.
I agree that if we want to test dimension this approach may be revised.

Reply:
I added dimensions test, but now I introduced again flakiness due to the `partition` dimension. Is `partition` an useful dimension to use?

By reading `Record.Partition` [docs](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo?utm_source=gopls#Record.Partition) it reads:
> For producing, this is left unset. This will be set by the client before the record is unbuffered

So I'm thinking this may not be an interesting producer attribute. What do you think?

Reply:
ah, what you're experiencing is caused by the fact that we have two partitions by default created by the helper `newClusterWithTopics`. I've opened a PR to change the signature of that function to require a partition number.

@axw I'm thinking it may be useful to see the producing throughput per partition rather than just on a per-topic aggregate. Do you think it's too much, or worth adding a dimension to it?